src/fastalloc/vregset.rs

```diff
@@ -0,0 +1,143 @@
+use core::fmt;
+
+use crate::ion::data_structures::VRegIndex;
+use crate::VReg;
+use alloc::vec;
+use alloc::vec::Vec;
+
+#[derive(Clone)]
+struct VRegNode {
+    next: VRegIndex,
+    prev: VRegIndex,
+    vreg: VReg,
+}
+
+// Using a doubly linked list here for fast insertion,
+// removal and iteration.
+pub struct VRegSet {
+    items: Vec<VRegNode>,
+    head: VRegIndex,
+}
+
+impl VRegSet {
+    pub fn with_capacity(num_vregs: usize) -> Self {
+        Self {
+            items: vec![
+                VRegNode {
+                    prev: VRegIndex::new(num_vregs),
+                    next: VRegIndex::new(num_vregs),
+                    vreg: VReg::invalid()
+                };
+                num_vregs + 1
+            ],
+            head: VRegIndex::new(num_vregs),
+        }
+    }
+
+    pub fn insert(&mut self, vreg: VReg) {
+        // Intentionally assuming that the set doesn't already
+        // contain `vreg`.
```

Comment:
Can we assert this? It looks like we always use `self.items[vreg.vreg()]` as the node for this entry -- can we check if its `vreg` field is `VReg::invalid()`, for example, and set it as such when we remove?